Add debounce to the status observers to reduce unnecessary CPU loops

afharo opened the pull request:
## Summary

Resolves #108950

During plugin registration, plugins can specify their own custom logic to declare the status. The status service was triggering an update of its observers for each registration, using unnecessary extra CPU loops.

This PR adds `debounceTime` to swallow any requests happening too close in time. 

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

lukeelmers commented on src/core/server/status/plugins_status.ts:
```diff
@@ -76,6 +76,7 @@ export class PluginsStatusService {
 
   public getDerivedStatus$(plugin: PluginName): Observable<ServiceStatus> {
     return this.update$.pipe(
+      debounceTime(25),
```
If we add a debounce here, do we still need it above? https://github.com/elastic/kibana/blob/524401973f1d0ac5403cd0fbb3ea82a63962a45a/src/core/server/status/plugins_status.ts#L72-L73

lukeelmers replied:
nit: Would be helpful to copy that comment here too for clarity

afharo replied:
I agree! Added a comment to the new debounceTime usage :)

lukeelmers reviewed (commented):
Overall makes sense to me, but to avoid regressions it would be nice if we could update the unit test to catch the edge case that wasn't covered by the existing debounce: https://github.com/elastic/kibana/blob/524401973f1d0ac5403cd0fbb3ea82a63962a45a/src/core/server/status/plugins_status.test.ts#L365

TinaHeiligers reviewed (approved):
With this change the setup to start time (locally) is less than a second:
```
[2021-08-17T10:54:48.282-07:00][INFO ][plugins-system.standard][]---Setting up [112] plugins: ...

//...setup logs

[2021-08-17T10:54:51.084-07:00][INFO ][plugins-system.standard][]---Starting [112] plugins: ...
```

One small suggestion is to use the same `const` for the debounce time in `getPluginStatuses ` and `getDerivedStatus$` but other than waiting for green CI, LGTM.

joshdover reviewed (approved):
Seems to work for me. +1 on improving the test coverage for this case.

TinaHeiligers reviewed (commented):
LGTM pending other ✅

joshdover commented:
@cachedout this may fix #107300 as well. Let's test it once it's been merged and available in a snapshot.

afharo commented:
Thank you all for your comments!

1. Added the test in [`1aa2639` (#108952)](https://github.com/elastic/kibana/pull/108952/commits/1aa2639a21070d56b65d7365ffd783dd1b940df9). 
2. I also had to update some previous tests since we now don't return the initial default `All dependencies are available` anymore thanks to the initial debounce
3. I also noticed that the `debounceTime` is only needed in one `this.update.pipe`. Updated the code for that and added a comment explaining why it's needed there.

lukeelmers reviewed (approved):
🚀

afharo commented on src/core/server/status/plugins_status.ts:
```diff
@@ -76,6 +76,7 @@ export class PluginsStatusService {
 
   public getDerivedStatus$(plugin: PluginName): Observable<ServiceStatus> {
     return this.update$.pipe(
+      debounceTime(25),
```
I removed it from `getPluginStatuses$`. I noticed it's not actually needed there.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/)
* Commit: 89243dc9d4d12ebe121341832df37caa17c1c4b6
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-108952/89243dc9d4d12ebe121341832df37caa17c1c4b6)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/execution/node/190/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/","number":"146274","commit":"89243dc9d4d12ebe121341832df37caa17c1c4b6"}]}
PIPELINE-->

afharo commented on src/core/server/status/plugins_status.test.ts:
```diff
@@ -247,7 +247,6 @@ describe('PluginStatusService', () => {
       subscription.unsubscribe();
 
       expect(statusUpdates).toEqual([
-        { a: { level: ServiceStatusLevels.available, summary: 'All dependencies are available' } },
```
No longer needed since we don't emit the initial transient status anymore.

elasticmachine commented:
Pinging @elastic/kibana-core (Team:Core)

cachedout commented:
> this may fix #107300 as well. Let's test it once it's been merged and available in a snapshot.

@joshdover Thanks! FWIW, I tried all day to replicate the slow-start Kibana problem by hand and I still can't get it to happen, but for whatever reason in the CI, we see it regularly. Hopefully this is the fix!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146316/)
* Commit: 1aa2639a21070d56b65d7365ffd783dd1b940df9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-108952/1aa2639a21070d56b65d7365ffd783dd1b940df9)
* [Documentation Changes](https://kibana_108952.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #146274](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/) failed 89243dc9d4d12ebe121341832df37caa17c1c4b6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/","number":"146274","commit":"89243dc9d4d12ebe121341832df37caa17c1c4b6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146316/","number":"146316","commit":"1aa2639a21070d56b65d7365ffd783dd1b940df9"}]}
PIPELINE-->

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/108998)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/108998">](https://github.com/elastic/kibana/pull/108998) |

This backport PR will be merged automatically after passing CI.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146332/)
* Commit: 1aa2639a21070d56b65d7365ffd783dd1b940df9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-108952/1aa2639a21070d56b65d7365ffd783dd1b940df9)
* [Documentation Changes](https://kibana_108952.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #146316](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146316/) succeeded 1aa2639a21070d56b65d7365ffd783dd1b940df9
* :broken_heart: [Build #146274](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/) failed 89243dc9d4d12ebe121341832df37caa17c1c4b6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146274/","number":"146274","commit":"89243dc9d4d12ebe121341832df37caa17c1c4b6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146316/","number":"146316","commit":"1aa2639a21070d56b65d7365ffd783dd1b940df9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/146332/","number":"146332","commit":"1aa2639a21070d56b65d7365ffd783dd1b940df9"}]}
PIPELINE-->

afharo commented:
Backporting to 7.14.1 as well: https://github.com/elastic/kibana/pull/109052

kobelb commented:
Great diagnosis @afharo <3
